runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -187,3 +187,29 @@ J9::ARM64::TreeEvaluator::monentEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    return targetRegister;
    }
 
+TR::Register *
+J9::ARM64::TreeEvaluator::arraylengthEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   // ldrw R, [B + contiguousSize]        
+   // cmp R, 0                            
+   // cselw R, [B + discontiguousSize]    
+   //
+   TR::Register *objectReg = cg->evaluate(node->getFirstChild());
+   TR::Register *lengthReg = cg->allocateRegister();
+   TR::Register *discontiguousLengthReg = cg->allocateRegister();
+
+   TR::MemoryReference *contiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfContiguousArraySizeField(), 4, cg);
+   TR::MemoryReference *discontiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfDiscontiguousArraySizeField(), 4, cg);   
```

Comment:
Shouldn't the discontiguous case memory size only be calculated if this is a discontiguous array? Same for the contiguous case? Can you check this beforehand?

Do we care about discontiguous arrays/arraylets at this point?

Reply:
Yes we care about the discontiguous arrays/arraylets. I am not sure whether we can check that beforehand or not because we are comparing the calculated memory size with 0 to differentiate between contiguous and discontiguous in the later instructions, and I have followed the structure of Power and X.